File: src/main/java/com/zl/pleasetweakwindows/RestorePointGuard.java
package com.zl.pleasetweakwindows;

import javafx.beans.property.BooleanProperty;
import javafx.scene.control.TextArea;
import javafx.stage.Stage;

public final class RestorePointGuard {
    private RestorePointGuard() {
    }

    private enum Decision {
        UNKNOWN,
        CREATED,
        SKIPPED
    }

    // Guarded by synchronized methods — no longer volatile
    private static Decision decision = Decision.UNKNOWN;
    private static final Object LOCK = new Object();

    public static void markCreated() {
        synchronized (LOCK) {
            decision = Decision.CREATED;
        }
    }

    public static void ensureRestorePoint(Stage owner,
                                          String scriptDirectory,
                                          TextArea logArea,
                                          BooleanProperty scriptsRunning,
                                          Runnable onProceed,
                                          Executor executor) {
        if (onProceed == null) {
            throw new IllegalArgumentException("onProceed callback must not be null");
        }

        synchronized (LOCK) {
            if (decision != Decision.UNKNOWN) {
                onProceed.run();
                return;
            }
        }

        DialogUtils.RestorePointDecision choice = DialogUtils.showRestorePointPrompt(owner);
        switch (choice) {
            case CREATE -> {
                String scriptPath = scriptDirectory + "create_restore_point.ps1";
                scriptsRunning.set(true);
                executor.runScript(scriptPath, logArea, (exitCode) -> {
                    if (exitCode == 0) {
                        synchronized (LOCK) {
                            decision = Decision.CREATED;
                        }
                    }
                    // Stay UNKNOWN on failure so user is prompted again next time
                    scriptsRunning.set(false);
                    onProceed.run();
                }, null);
            }
            case SKIP -> {
                synchronized (LOCK) {
                    decision = Decision.SKIPPED;
                }
                onProceed.run();
            }
            case CANCEL -> {
                // user cancelled
            }
        }
    }
}

File: PleaseTweakWindows.Avalonia/src/PleaseTweakWindows/Services/RestorePointGuard.cs
using Microsoft.Extensions.Logging;

namespace PleaseTweakWindows.Services;

public sealed class RestorePointGuard : IRestorePointGuard
{
    private enum Decision { Unknown, Created, Skipped }

    private readonly IDialogService _dialogService;
    private readonly IScriptExecutor _executor;
    private readonly ILogger<RestorePointGuard> _logger;
    private readonly object _lock = new();
    private Decision _decision = Decision.Unknown;

    public RestorePointGuard(IDialogService dialogService, IScriptExecutor executor, ILoggerFactory loggerFactory)
    {
        _dialogService = dialogService;
        _executor = executor;
        _logger = loggerFactory.CreateLogger<RestorePointGuard>();
    }

    public void MarkCreated()
    {
        lock (_lock)
        {
            _decision = Decision.Created;
        }
    }

    /// <summary>
    /// Returns true if the user wants to proceed, false if cancelled.
    /// </summary>
    public async Task<bool> EnsureRestorePointAsync(string scriptDirectory, Action<string>? onOutput, CancellationToken cancellationToken = default)
    {
        lock (_lock)
        {
            if (_decision != Decision.Unknown)
                return true;
        }

        var choice = await _dialogService.ShowRestorePointPromptAsync();

        switch (choice)
        {
            case RestorePointDecision.Create:
                var scriptPath = Path.Combine(scriptDirectory, "create_restore_point.ps1");
                var exitCode = await _executor.RunScriptAsync(scriptPath, null, onOutput, cancellationToken);
                if (exitCode == 0)
                {
                    lock (_lock) { _decision = Decision.Created; }
                }
                // Stay Unknown on failure so user is prompted again next time
                return true;

            case RestorePointDecision.Skip:
                lock (_lock) { _decision = Decision.Skipped; }
                return true;

            case RestorePointDecision.Cancel:
            default:
                return false;
        }
    }
}
